asserts/pool_test.go

```diff
@@ -691,6 +878,396 @@ func (s *poolSuite) TestUpdate(c *C) {
+	assertstest.AddMany(s.db, s.hub.StoreAccountKey(""), s.seq1_1111r5)
+
+	pool := asserts.NewPool(s.db, 64)
+
+	atseq := &asserts.AtSequence{
+		Type:        s.seq1_1111r5.Type(),
+		SequenceKey: []string{"1111"},
+		Sequence:    s.seq1_1111r5.Sequence(),
+		Revision:    s.seq1_1111r5.Revision(),
+	}
+	err := pool.AddSequenceToUpdate(atseq, "for_one") // group num: 0
+	c.Assert(err, IsNil)
+
+	toResolve, toResolveSeq, err := pool.ToResolve()
+	c.Assert(err, IsNil)
+	c.Check(toResolve, HasLen, 0)
+	c.Check(toResolveSeq, DeepEquals, map[asserts.Grouping][]*asserts.AtSequence{
+		asserts.MakePoolGrouping(0): {
+			&asserts.AtSequence{
+				Type:        s.seq1_1111r5.Type(),
+				SequenceKey: []string{"1111"},
+				Sequence:    1,
+				Revision:    5,
+			}},
+	})
+
+	c.Check(pool.Err("for_one"), IsNil)
+
+	// update with same assertion
+	ok, err := pool.Add(s.seq1_1111r5, asserts.MakePoolGrouping(0))
+	c.Assert(err, IsNil)
+	c.Assert(ok, Equals, true)
+
+	toResolve, toResolveSeq, err = pool.ToResolve()
+	c.Assert(err, IsNil)
+	c.Check(toResolve, HasLen, 0)
+	c.Check(toResolveSeq, HasLen, 0)
+	c.Check(pool.Err("for_one"), IsNil)
+
+	err = pool.CommitTo(s.db)
+	c.Check(err, IsNil)
+	c.Assert(pool.Err("for_one"), IsNil)
+
+	// sequence point 1, revision 5 is still in the database.
+	_, err = s.seq1_1111r5.Ref().Resolve(s.db.Find)
+	c.Assert(err, IsNil)
+}
+
+func (s *poolSuite) TestUpdateSeqFormingPinnedNewerSequenceSameRevisionNoop(c *C) {
+	assertstest.AddMany(s.db, s.hub.StoreAccountKey(""), s.seq1_1111r5)
+
+	pool := asserts.NewPool(s.db, 64)
+
+	atseq := &asserts.AtSequence{
+		Type:        s.seq1_1111r5.Type(),
+		SequenceKey: []string{"1111"},
+		Sequence:    s.seq1_1111r5.Sequence(),
+		Revision:    s.seq1_1111r5.Revision(),
+		Pinned:      true,
+	}
+	err := pool.AddSequenceToUpdate(atseq, "for_one") // group num: 0
+	c.Assert(err, IsNil)
+
+	toResolve, toResolveSeq, err := pool.ToResolve()
+	c.Assert(err, IsNil)
+	c.Check(toResolve, HasLen, 0)
+	c.Check(toResolveSeq, DeepEquals, map[asserts.Grouping][]*asserts.AtSequence{
+		asserts.MakePoolGrouping(0): {
+			&asserts.AtSequence{
+				Type:        s.seq1_1111r5.Type(),
+				SequenceKey: []string{"1111"},
+				Sequence:    1,
+				Revision:    5,
+				Pinned:      true,
+			}},
+	})
+
+	c.Check(pool.Err("for_one"), IsNil)
+
+	// resolve
+	ok, err := pool.Add(s.seq3_1111r5, asserts.MakePoolGrouping(0))
+	c.Assert(err, IsNil)
+	c.Assert(ok, Equals, true)
+
+	toResolve, toResolveSeq, err = pool.ToResolve()
+	c.Assert(err, IsNil)
+	c.Check(toResolve, HasLen, 0)
+	c.Check(toResolveSeq, HasLen, 0)
+	c.Check(pool.Err("for_one"), IsNil)
+
+	err = pool.CommitTo(s.db)
+	c.Check(err, IsNil)
+	c.Assert(pool.Err("for_one"), IsNil)
+
+	// sequence point 1, revision 5 is still the latest.
+	_, err = s.seq1_1111r5.Ref().Resolve(s.db.Find)
+	c.Assert(err, IsNil)
+
+	// and sequence point 3 revision 5 wasn't added to asserts database.
```

Comment:
just to clarify this, the reason it wasn't added is because we pinned to sequence point 1 correct? followup question is what exactly is pinned? would we say that the pool was pinned to the sequence point 1? or would it be more accurate to say that the sequence forming assertion was pinned to the pool? it's a bit difficult to follow conceptually the direction of the "pinning"

Reply:
It's about a validation set assertion being pinned to the given sequence point (by the user). We tell pool about assertions we want to update or resolve, pool only "manages" the updating process. This test is about making sure that if validation set assertion was pinned at the given sequence point, it won't get updated with a newer one (coming from the store). To be clear, if there was a newer assertion with same sequence number but higher revision, then it would get updated (TestUpdateSeqFormingPinnedSameSequenceNewerRevision is about that case).

Reply:
right got it, thanks for explaining